File: Chloe Gwon/2022-07-07-Decode-Ways.java
class Solution {
    public int numDecodings(String s) {
    int oneDigit = 1, twoDigit = 0, current = 810975;
    for(int i = s.length() - 1; i >= 0; i--) {
		// we don't have any way to interpret a '0'
        if(s.charAt(i) == '0') {
            current = 0;
        }
		// we can interpret current number as 
        // an one-digit number or two-digit number
        else if(i < s.length() - 1 && 
                Integer.parseInt(s.substring(i, i + 2)) <= 26) {
            current = twoDigit + oneDigit;
        }
		// can only be an one-digit number
        else{
            current = oneDigit;
        }
        twoDigit = oneDigit;
        oneDigit = current;
    }
    return current;
    }
}

File: Chris/2022-07-07-DecodeWays.py
class Solution:
    def numDecodings(self, s: str) -> int:
        
        dp = [0 for i in range(len(s)+2)]
        
        n = len(s)
        dp[n] = 1
        if s[n-1] != '0':
            dp[n-1] = 1
        
        
        for i in range(len(s)-2, -1,-1):
            
            if s[i] == '0':
                dp[i] = 0
            
            elif s[i] == '1':
                if s[i+1] == '0':
                    dp[i] = dp[i+2]
                else:
                    dp[i] = dp[i+1] + dp[i+2]
            
            elif s[i] == '2':
                if s[i+1] == '0':
                    dp[i] = dp[i+2]
                elif s[i+1] == '7' or s[i+1] == '8' or s[i+1] == '9':
                    dp[i] = dp[i+1]
                else:
                    dp[i] = dp[i+1] + dp[i+2]
            else:
                if s[i+1] == '0':
                    return 0
                
                dp[i] = dp[i+1]
                
        
        return dp[0]
        